File: app/src/main/java/com/beautifulquran/ui/theme/SettingsNuqta.kt
package com.beautifulquran.ui.theme

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Tune
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.graphics.Paint
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.graphics.drawscope.translate
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.graphics.vector.rememberVectorPainter
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.roundToInt
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

/** The paper stack's turn toward Settings, as the settings nuqta reads it. */
@Stable
class SettingsApproach(
    /**
     * How far the stack has turned from the sheet beneath Settings toward
     * Settings itself, 0..1. Read it only while drawing or in snapshot flows:
     * it changes every frame of a swipe.
     */
    val progress: () -> Float,
    /** Whether a finger is dragging the stack right now. */
    val dragging: () -> Boolean,
    /** The share of a drag past which letting go completes the turn. */
    val commitAt: Float,
)

val LocalSettingsApproach = staticCompositionLocalOf {
    SettingsApproach(progress = { 0f }, dragging = { false }, commitAt = 1f)
}

/**
 * Lets the settings button ink its nuqta while it is held — pass
 * [interactions] to its clickable — and the instant it is tapped.
 */
@Stable
class SettingsNuqtaState {
    /** The button's presses: holding it inks the nuqta as its pressed state. */
    val interactions = MutableInteractionSource()

    /** Tapped, and the turn it starts has not yet settled. */
    internal var pressed by mutableStateOf(false)

    /** Touch the pen down: the nuqta spreads on its own select clock. */
    fun drop() {
        pressed = true
    }
}

@Composable
fun rememberSettingsNuqtaState(): SettingsNuqtaState = remember { SettingsNuqtaState() }

/**
 * The settings glyph with a nuqta of ink behind it, so the reader sees where
 * a page turn will land. The moment the button is held, or the stack starts
 * to turn toward Settings — a swipe or a tap ([SettingsNuqtaState.drop]) —
 * the drop spreads on the same clock as a chosen row.
 *
 * The drag then stretches it like a rubber band in two stages: once the
 * drag passes the point where letting go completes the turn, the drop
 * bounces up a stage with a haptic tick. Holding the button shows the band
 * at its fullest. If the press or turn falls back short of Settings the ink
 * dries in place, its layers fading one after another. The glyph takes the contrasting colour only
 * where the ink covers it.
 */
@Composable
fun SettingsNuqtaIcon(
    state: SettingsNuqtaState,
    contentDescription: String?,
    tint: Color,
    modifier: Modifier = Modifier,
    iconSize: Dp = 26.dp,
    params: NuqtaParams = LocalNuqtaParams.current,
) {
    val turn = LocalSettingsApproach.current
    val approach = turn.progress
    val held by state.interactions.collectIsPressedAsState()
    val view = LocalView.current
    val spread = remember { Animatable(0f) }
    val presence = remember { Animatable(0f) }
    // The band's second stage, on a bouncy spring: 0 before the commit
    // point, 1 past it.
    val stage = remember { Animatable(0f) }
    // Holding the button pulls the band to its fullest.
    val pull = remember { Animatable(0f) }
    LaunchedEffect(turn, state) {
        // A tap's turn carries the ink until it lands on Settings, or until
        // a finger takes the page over.
        snapshotFlow { approach() >= 1f - SettingsTurnEpsilon || turn.dragging() }.collect { done ->
            if (done) state.pressed = false
        }
    }
    LaunchedEffect(turn, state) {
        snapshotFlow { held || state.pressed || approach() >= turn.commitAt }
            .distinctUntilChanged()
            .collectLatest { committed ->
                if (committed && turn.dragging()) view.paperSelectHaptic()
                stage.animateTo(
                    if (committed) 1f else 0f,
                    spring(dampingRatio = 0.38f, stiffness = Spring.StiffnessMedium),
                )
            }
    }
    LaunchedEffect(state) {
        snapshotFlow { held }.collectLatest { holding ->
            pull.animateTo(
                if (holding) settingsNuqtaStretch(1f) else 0f,
                spring(dampingRatio = 0.5f, stiffness = Spring.StiffnessMediumLow),
            )
        }
    }
    LaunchedEffect(turn, state, params) {
        snapshotFlow { approach() > SettingsTurnEpsilon || state.pressed || held }
            .distinctUntilChanged()
            .collectLatest { inked ->
                if (inked) {
                    if (presence.value <= 0f) spread.snapTo(0f)
                    // Turned again while it dries: the colour floods back into
                    // the drop as it is, and it carries on spreading.
                    coroutineScope {
                        launch {
                            presence.animateTo(1f, tween(SettingsNuqtaRewetMs, easing = LinearEasing))
                        }
                        spread.animateTo(
                            1f,
                            tween((params.spreadMs * (1f - spread.value)).roundToInt(), easing = LinearEasing),
                        )
                    }
                } else {
                    // A press or turn let go short of Settings dries in place:
                    // the drop keeps its shape while its layers fade one after
                    // another, the coat paling toward the paper as it goes.
                    presence.animateTo(0f, tween(SettingsNuqtaDryMs, easing = LinearEasing))
                    spread.snapTo(0f)
                }
            }
    }
    val fingers = remember(params.seed) { nuqtaFingers(params.seed) }
    val painter = rememberVectorPainter(Icons.Rounded.Tune)
    val ink = MaterialTheme.colorScheme.primary
    val paper = MaterialTheme.colorScheme.background
    val inked = MaterialTheme.colorScheme.onPrimary

    Canvas(
        modifier
            .size(iconSize * SettingsNuqtaScale)
            .semantics { if (contentDescription != null) this.contentDescription = contentDescription },
    ) {
        val t = spread.value
        val lift = presence.value
        val glyphPx = iconSize.toPx()
        fun glyph(color: Color) = translate(
            left = (size.width - glyphPx) / 2f,
            top = (size.height - glyphPx) / 2f,
        ) {
            with(painter) { draw(Size(glyphPx, glyphPx), colorFilter = ColorFilter.tint(color)) }
        }
        if (t <= 0f || lift <= 0f) {
            glyph(tint)
            return@Canvas
        }
        val swell = 1f + maxOf(settingsNuqtaStretch(approach()), pull.value) +
            SettingsNuqtaStageSwell * stage.value
        // A drying coat is a thinner one: its pigment pales toward the paper.
        val coat = lerp(ink, lerp(ink, paper, SettingsNuqtaDriedTint), 1f - lift)
        scale(swell) { drawInkNuqta(t, 1f, params, fingers, coat, Color.Transparent, dry = 1f - lift) }
        glyph(tint)
        // Where the ink lies, the glyph turns to paper exactly as far as the
        // ink under it is dense: a contrasting copy, masked by the same drop.
        drawIntoCanvas { canvas ->
            val bounds = Rect(Offset.Zero, size)
            canvas.saveLayer(bounds, Paint())
            glyph(inked)
            // The mask is its own layer so that, composited DstIn, its bare
            // paper clears the copy too — not only where ink was drawn.
            canvas.saveLayer(bounds, Paint().apply { blendMode = BlendMode.DstIn })
            scale(swell) {
                drawInkNuqta(t, 1f, params, fingers, Color.Black, Color.Transparent, dry = 1f - lift)
            }
            canvas.restore()
            canvas.restore()
        }
    }
}

/** The nuqta's size as a multiple of the glyph it sits behind. */
private const val SettingsNuqtaScale = 1.6f

/** A drop let go dries in place over this long, layer by layer. */
private const val SettingsNuqtaDryMs = 560

/** Turned again mid-dry, the colour floods back over this long. */
private const val SettingsNuqtaRewetMs = 180

/** How far toward the paper the last of a drying coat has paled. */
private const val SettingsNuqtaDriedTint = 0.55f

/** The extra swell the band bounces up by as the drag passes the commit point. */
private const val SettingsNuqtaStageSwell = 0.14f

/** How far the stack must move before it counts as turning toward Settings. */
private const val SettingsTurnEpsilon = 0.001f

/** The swell the rubber band tends toward if the turn could run on forever. */
private const val SettingsNuqtaStretchLimit = 0.8f

/** How stiff the band is: its pull, per share of the turn, as it starts to stretch. */
private const val SettingsNuqtaStretchStiffness = 1f

/**
 * The swell past full size at [approach]. The drop is a rubber band: it
 * grows with the drag all the way to Settings, but each step buys less than
 * the one before, like tension building before it snaps back.
 */
internal fun settingsNuqtaStretch(approach: Float): Float {
    val over = approach.coerceIn(0f, 1f)
    return SettingsNuqtaStretchLimit * (1f - 1f / (1f + SettingsNuqtaStretchStiffness * over))
}

File: web/src/ui/theme/settingsNuqta.ts
/**
 * The swell past full size at [approach]. Android `settingsNuqtaStretch`:
 * a rubber band — each step of the drag buys less than the one before.
 */
export function settingsNuqtaStretch(approach: number): number {
  const over = Math.min(1, Math.max(0, approach))
  const limit = 0.8
  const stiffness = 1
  return limit * (1 - 1 / (1 + stiffness * over))
}

export const SETTINGS_NUQTA_SCALE = 1.6
export const SETTINGS_NUQTA_DRY_MS = 500
export const SETTINGS_NUQTA_REWET_MS = 180
export const SETTINGS_NUQTA_DRIED_TINT = 0.55
export const SETTINGS_NUQTA_STAGE_SWELL = 0.14
export const SETTINGS_TURN_EPSILON = 0.001
/** Android `STACK_PAGE_TURN_THRESHOLD` — letting go past this completes the turn. */
export const SETTINGS_COMMIT_AT = 0.15

const STIFFNESS_MEDIUM = 1500
const STIFFNESS_MEDIUM_LOW = 400

/** Semi-implicit spring, unit mass. [dt] is seconds. */
export function springStep(
  value: number,
  velocity: number,
  target: number,
  dt: number,
  stiffness: number,
  dampingRatio: number,
): { value: number; velocity: number } {
  const step = Math.min(Math.max(dt, 0), 0.032)
  const omega = Math.sqrt(stiffness)
  const damping = 2 * dampingRatio * omega
  const acceleration = -stiffness * (value - target) - damping * velocity
  const nextVelocity = velocity + acceleration * step
  return { value: value + nextVelocity * step, velocity: nextVelocity }
}

export type SettingsNuqtaPose = {
  spread: number
  /** 1 wet, 0 dried out. Android `presence`. */
  presence: number
  stage: number
  pull: number
}

/**
 * One settings glyph's nuqta. Spread follows the select clock; the rubber
 * band follows the sheet; a press or a short turn dries the coats in place.
 */
export class SettingsNuqtaMachine {
  spread = 0
  presence = 0
  stage = 0
  pull = 0
  pressed = false
  held = false

  private stageV = 0
  private pullV = 0
  private spreading = false
  private spreadElapsed = 0
  private spreadDur = 0
  private presenceElapsed = 0
  private presenceDur = 0
  private presenceFrom = 0
  private presenceTo = 0
  private presenceRunning = false

  pose(): SettingsNuqtaPose {
    return {
      spread: this.spread,
      presence: this.presence,
      stage: this.stage,
      pull: this.pull,
    }
  }

  /**
   * True only while something is still moving. A drop that has settled on
   * Settings stays painted and does not keep a frame loop alive.
   */
  busy(): boolean {
    const drying = this.presence > 0.001 && this.presence < 0.999
    const spreading = this.spread > 0.001 && this.spread < 0.999 && this.presence > 0
    return (
      this.pressed ||
      this.held ||
      this.spreading ||
      this.presenceRunning ||
      drying ||
      spreading ||
      Math.abs(this.stageV) > 0.02 ||
      Math.abs(this.pullV) > 0.02
    )
  }

  /**
   * [spreadMs] is the shipped nuqta's touch-to-settle clock.
   * [approach] is the stack's turn into Settings, 0..1.
   */
  step(dtMs: number, approach: number, spreadMs: number): void {
    const dt = dtMs / 1000
    const inked = approach > SETTINGS_TURN_EPSILON || this.pressed || this.held
    if (inked && !this.spreading && this.spread < 1) {
      // A fresh drop starts empty. Turned again while it dries, the colour
      // floods back into the drop as it stands.
      if (this.presence <= 0) this.spread = 0
      this.spreading = true
      this.spreadElapsed = 0
      this.spreadDur = Math.max(1, spreadMs * (1 - this.spread))
      this.startPresence(1, SETTINGS_NUQTA_REWET_MS)
    } else if (!inked && (this.spreading || this.presence > 0)) {
      this.spreading = false
      if (this.presenceTo !== 0 || !this.presenceRunning) {
        this.startPresence(0, SETTINGS_NUQTA_DRY_MS)
      }
    }
    if (approach >= 1 - SETTINGS_TURN_EPSILON) this.pressed = false

    if (this.spreading) {
      this.spreadElapsed += dtMs
      const u = Math.min(1, this.spreadElapsed / this.spreadDur)
      const from = 1 - this.spreadDur / spreadMs
      this.spread = from + (1 - from) * u
      if (u >= 1) {
        this.spread = 1
        this.spreading = false
      }
    }

    this.advancePresence(dtMs)

    const committed = this.held || this.pressed || approach >= SETTINGS_COMMIT_AT
    const stageTarget = committed ? 1 : 0
    const stage = springStep(this.stage, this.stageV, stageTarget, dt, STIFFNESS_MEDIUM, 0.38)
    if (Math.abs(stage.value - stageTarget) < 0.001 && Math.abs(stage.velocity) < 0.02) {
      this.stage = stageTarget
      this.stageV = 0
    } else {
      this.stage = stage.value
      this.stageV = stage.velocity
    }

    const pullTarget = this.held ? settingsNuqtaStretch(1) : 0
    const pull = springStep(this.pull, this.pullV, pullTarget, dt, STIFFNESS_MEDIUM_LOW, 0.5)
    if (Math.abs(pull.value - pullTarget) < 0.001 && Math.abs(pull.velocity) < 0.02) {
      this.pull = pullTarget
      this.pullV = 0
    } else {
      this.pull = pull.value
      this.pullV = pull.velocity
    }
  }

  private startPresence(to: number, dur: number): void {
    this.presenceFrom = this.presence
    this.presenceTo = to
    this.presenceElapsed = 0
    this.presenceDur = dur
    this.presenceRunning = true
  }

  private advancePresence(dtMs: number): void {
    if (!this.presenceRunning) return
    this.presenceElapsed += dtMs
    const u = Math.min(1, this.presenceElapsed / Math.max(1, this.presenceDur))
    this.presence = this.presenceFrom + (this.presenceTo - this.presenceFrom) * u
    if (u >= 1) {
      this.presence = this.presenceTo
      this.presenceRunning = false
      if (this.presence <= 0) this.spread = 0
    }
  }
}
